src/vsr/replica.zig

```diff
@@ -3522,11 +3522,19 @@ pub fn ReplicaType(
                         nacks.set(i);
                     }
 
+                    // Presence bit: We have the header, and the prepare is on disk,
+                    // is being written to disk, or is cached in memory.
                     if (journal_header != null and journal_header.?.checksum == header.checksum and
                         !faulty)
                     {
-                        maybe(nacks.isSet(i));
-                        present.set(i);
+                        if ((self.journal.prepare_inhabited[slot.index] and
+                            self.journal.prepare_checksums[slot.index] == header.checksum) or
+                            self.journal.writing(header.op, header.checksum) or
+                            self.pipeline_prepare_by_op_and_checksum(header.op, header.checksum) != null)
```

Comment:
This is interesting. I _think_ we want this condition to be more or less symmetric to that in `on_request_prepare`. But we don't have `self.journal.writing(header.op, header.checksum)` case in `on_request_prepare`; there, we only try cache and disk, we don't try in-flight writes. 

I am wondering though, can it be the case that the prepare is being written, but is absent from the cache? Maybe `self.journal.writing(header.op, header.checksum)` condition is fully covered by `self.pipeline_prepare_by_op_and_checksum(header.op, header.checksum) != null` ?

Reply:
Cache is updated in `write_prepare()`:

```zig

            // Criteria for caching:
            // - The primary does not update the cache since it is (or will be) reconstructing its
            //   pipeline.
            // - Cache uncommitted ops, since it will avoid a WAL read in the common case.
            if (self.pipeline == .cache and
                self.replica != self.primary_index(self.view) and
                self.commit_min < message.header.op)
            {
                const prepare_evicted = self.pipeline.cache.insert(message.ref());
                if (prepare_evicted) |m| self.message_bus.unref(m);
            }
```

So I think there _are_ cases where we write a prepare which is not in our cache.